README
NAME
    Geo::Proj4 - Wrap the powerful PROJ.4 cartographic projections library

SYNOPSIS
      use Geo::Proj4;

      my $proj = Geo::Proj4->new(
            proj => "merc", ellps => "clrk66", lon_0 => -96 );

      my ($x, $y) = $proj->forward($lat, $lon);

      my ($lat, $lon) = $proj->inverse($x, $y);

DESCRIPTION
    This perl library converts geodetic latitude and longitude into an
    enormous variety of cartographic projections and back. Geo::Proj4 uses
    XS to wrap the PROJ.4 cartographic projections library. You will need to
    have the PROJ.4 library installed in order to build and use this module.
    You can get source code and binaries for the PROJ.4 library from its
    home page at http://www.remotesensing.org/proj/. See the pj_init(3)
    manpage for details on the internals.

METHODS
    new( %args )
          my $proj = Geo::Proj4->new(
                    proj => "merc", ellps => "clrk66", lon_0 => -96 );

        The contructor accepts a hash of parameters that will be passed on
        to the PROJ.4 library. You must supply a "proj" parameter
        identifying the target projection. Specify boolean parameters (e.g.
        the south parameter to the UTM projection) with a matching value of
        undef.

        Covering all the possible projections and their arguments in PROJ.4
        is well beyond the scope of this document. However, the cs2cs
        utility that ships with PROJ.4 will list the projections it knows
        about by running cs2cs -lp, the ellipsoid models it knows with the
        -le parameter, the units it knows about with -lu, and the geodetic
        datums it knows with -ld. Read the cs2cs(1) manpage for more
        details.

        Alternately, you can read the PROJ.4 documentation, which can be
        found on the project's homepage. There are links to PDFs, text
        documentation, a FAQ, and more.

    forward( $lat, $lon )
          my ($x, $y) = $proj->forward($lat, $lon);

        Perform a forward projection from latitude and longitude to the
        cartographic projection represented by $proj. Latitude and longitude
        are assumed to be in degrees, with latitude south of the Equator and
        longitude west of the Prime Meridian given with negative values. $x
        and $y are typically returned in meters, or whatever units are
        relevant to the given projection.

        If PROJ.4 encounters an error, forward() will return undef for both
        values.

    inverse( $x, $y )
          my ($lat, $lon) = $proj->inverse($x, $y);

        Perform an inverse projection from the cartographic projection
        represented by $proj back into latitude and longitude. Units and
        error conditions are as described above.

EXAMPLES
      To Convert from Lat/Long to UTM:
      #!/usr/bin/perl

      use strict;
      use Geo::Proj4;

      my $proj = Geo::Proj4->new( proj => "utm", zone => 10 );
      my ($x, $y) = $proj->forward(38.40342, -122.81856);
      print "conversion to UTM: y is  $y\n";
      print "conversion to UTM: x is  $x\n";

      my ($lat, $long) = $proj->inverse($x, $y);
      print "inverse conversion: lat is $lat \n" ;
      print "inverse conversion: long is $long \n" ;

ERRATA, BUGS, TODO, ETC.
    One common source of errors is that latitude and longitude are swapped,
    or that the values have the wrong sign. Make sure you give negative
    values for south latitude and west longitude.

    PROJ.4 offers a "pj_transform()" function that would be really cool to
    add to this module, but... PROJ.4 expects spherical coordinates (i.e.
    lat and long) in radians, whereas most ordinary people think of lat and
    long in degrees. Ordinarily, Geo::Proj4 takes care of this for you
    transparently, but in the case of the proposed "transform()" method, the
    list of points to be transformed between two cartographic projections
    might be spherical or planar in zero, one, or both of the input and
    output lists. We could supply helper functions to allow a developer to
    perform the degree-to-radian and reverse as needed before or after the
    call to "transform()", but... would anyone use this, or would you just
    rely on cs2cs?

    Passing undef as the value to a boolean parameter is an ugly interface.

    Needs more tests! Especially of border cases.

    Needs more samples in the samples directory!

SEE ALSO
    PROJ.4 home page: http://www.remotesensing.org/proj/

    Mapping Hacks home page: http://www.mappinghacks.com

    proj(1), cs2cs(1), pj_init(3), Geo::Dymaxion(3pm),
    Geo::Coordinates::UTM(3pm)

AUTHORS
    Schuyler Erle <schuyler@nocat.net>

    Rich Gibson <rich@nocat.net>

EFFUSIVE THANKS TO...
    Frank Warmerdam, maintainer of PROJ.4 and all around most excellent
    person.

    Gerald Evenden. 'Essentially all work (on PROJ.4) was done by Gerald...'

COPYRIGHT
    Copyright (C) 2004 by Schuyler Erle and Rich Gibson

LICENSE
    This library is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself, either Perl version 5.8.3 or, at
    your option, any later version of Perl 5 you may have available.



Changes
Revision history for Perl extension Geo::Proj4.

version 1.05: Tue Jan 21 14:32:06 CET 2014

	Improvements:
	- change documentation style.

version 1.04: Fri Sep  7 12:54:23 CEST 2012

	Fixes:

	- proj4 version 4.8.0 does not install project.h anymore. Therefore,
	  we carry our own now.  Flagged by [Jeff Huber]

version 1.03: Wed Feb  8 19:57:47 CET 2012

	Fixes:

	- \ missing in pattern match of ::projection()
	  rt.cpan.org#74774 [Heiko Klein]

version 1.02: Wed Nov 23 09:24:16 CET 2011

	Fixes:

	- with Perl 14.2, the XS compile flags changed slightly causing
	  a Dynaloader error.
          rt.cpan.org#72626 [Dirk Stöcker]

	Improvements:

	- a little better installation under Windows [André Joost]

version 1.01: Tue Jan 27 16:41:47 CET 2009

	- show ->new(init => ) example in main man-page.
	  Contributed by [Michael R. Davis]

	- improved output of examples/test_list.pl

	- do not list latlong/longlat/geocent as projections,
	  because "proj -l" doesn't either.

	- tested with 4.6.1

version 1.00: Thu Oct  2 23:27:31 CEST 2008

	- XS should use Safefree(), not free().  Marvellous fix
	  by rt.cpan.org#39749 [Dan Stahlke]

version 0.99: Fri Jun  8 17:08:40 CEST 2007

	- EPGS -> EPSG in example script [Michael R. Davis]

	- added t/pod.t

version 0.98: Fri Mar  9 14:16:09 CET 2007

	- moved eg/ to examples/, probably easier to find.

	- added examples/use_epsg.pl, and use of epgs files in synopsis.
	  Thanks to [Michael R. Davis]  Translated POD in script to text,
	  to avoid search.cpan.org to index it as package documentation.

	- converted production of manual pages to use oodist, which
	  eliminated the need for the mkdist, mkdoc, version files,
	  but added PODTAIL.txt and text to Makefile.PL

	- renamed test_list into examples/test_list.pl

	- removed use of ppport.h, which is probably not used.

	- added documentation for the two error messages.

version 0.97: Sat Nov  4 12:14:22 CET 2006
	- Avoid double free() on DESTROY by simply forgetting to free
	  the memory.  Of course this leaks, which should be fixed in
	  next release.

version 0.96: Mon Dec 12 11:58:18 CET 2005
	- added projection() and datum()
	- renamed or datum(), ellipsoid(), and unit() into ....Info()

version 0.95: Fri Nov 25 17:27:21 CET 2005  (public release)
	- minor doc improvements

version 0.12: Sat Nov 19 10:37:45 CET 2005  (internal release)
	- new maintainer: Mark Overmeer
	- converted ExtUtils::MakeMaker in Makefile.PL
	- moved Proj4.pm to lib/Geo/Proj4
	- create manual page with OODoc, to be able to integrate it with
	  Geo::Point docs.
	- So many extensions!!!!

version 0.11: Sat Mar 27 15:52:42 PST 2004
	- Fixed build errors in typemap with more current versions of gcc.
	- Switched Makefile.PL to use Module::Install.

version 0.10: Sat Mar 27 03:44:33 PST 2004
	- Initial cut of Geo::Proj4.


examples/test_list.pl
#!/usr/bin/perl -w
# when run, it will show all definitions.
use warnings;
use strict;

use lib qw[blib/lib blib/arch ../blib/lib ../blib/arch];

use Geo::Proj4;

print "Ellisoids:\n";
my @ells = Geo::Proj4->listEllipsoids;
foreach my $ell (@ells)
{   my $def = Geo::Proj4->ellipsoidInfo($ell);
    print "  $ell: $def->{name}\n"
        , "      ellipses: $def->{ell}"
        , ", major: $def->{major}\n";
}

print "\nUnits:\n";
my @units = Geo::Proj4->listUnits;
foreach my $unit (@units)
{   my $def = Geo::Proj4->unitInfo($unit);
    printf "  %6s: %12.6f meter; %s\n", $unit, $def->{to_meter}, $def->{name};
}

print "\nDatums:\n";
my @datums = Geo::Proj4->listDatums;
foreach my $datum (@datums)
{   my $def = Geo::Proj4->datumInfo($datum);
    print "  $datum: ", ($def->{comments}||''), "\n"
        , "      $def->{definition}\n";
}

print "\nProjections:\n";
my @types = Geo::Proj4->listTypes;
foreach my $type (@types)
{   my $def = Geo::Proj4->typeInfo($type);
    my $descr = $def->{description} || '';
    $descr  =~ s/\s*\z//;
    $descr .= "\n(has inverse)" if $def->{has_inverse};
    $descr  =~ s/\n\s*/\n      /g;
    print "  $type: $descr\n";
}


Makefile.PL
use ExtUtils::MakeMaker;
                                                                                
require 5.006;

# increased per release
my $relversion = '1.05';

my $libversion = '4.4.9';

# Should also work when you do not have FWTools installed...
my $FWTools    = $ENV{GEOPROJ_FWTOOLS_DIR}
              || $ENV{FWTOOLS_DIR}           # windows
              || '.';

# This is part is very UNIX/Linux centric.  Please help me with the
# required windows tests.
my $FWbin      = "$FWTools/bin_safe";
my $cs2cs      = -x "$FWbin/cs2cs" ? "$FWbin/cs2cs" : 'cs2cs';

open CS, "$cs2cs 2>&1 |"
   or die "ERROR: proj library not found, where is cs2cs?\n";

my $got = <CS>;

$got =~ m/\ ([\d.]+)\,/
    or die "ERROR: proj library version not known\n";

my $version = $1;
$version ge $libversion
    or die "ERROR: libproj too old, found $version required is $libversion\n";

unlink 'projects.h';
my $own_projects = "src/$version/projects.h";
if(-f $own_projects)
{   symlink $own_projects, 'projects.h'
        or die "ERROR: cannot install $own_projects: $!\n";
    warn "installed $own_projects\n";
}

WriteMakefile
 ( NAME     => 'Geo::Proj4'
 , VERSION  => $relversion
 , AUTHOR   => 'Mark Overmeer'
 , ABSTRACT => 'Proj4 library for carthographic projections'
 , INC      => "-I. -I$FWTools/include"
 , LIBS     => [ "-L$FWTools/lib -lproj" ]
 , LICENSE  => 'perl'
 );

### used by oodist during production of this distribution
sub MY::postamble { <<'__POSTAMBLE' }

# for DIST, see PODTAIL.txt as well
LICENSE		= artistic
RAWDIR          = ../public_html/geoproj4/raw
DISTDIR         = ../public_html/geoproj4/source
__POSTAMBLE


examples/use_epsg.pl
#!/usr/bin/perl -w
# Example contributed by Michael R. Davis, see after __END__

use strict;
use warnings;
use Geo::Proj4 ();

my $epsg = 26985;
my $proj = Geo::Proj4->new(init => "epsg:$epsg")
  or die "cannot use EPGS 26985: ",Geo::Proj4->error, "\n";

my ($x, $y) = (401717.80, 130013.88);
my ($lat, $lon) = $proj->inverse($x, $y);
print "  x: $x\n  y: $y\nlat: $lat\nlon: $lon\n";

__END__

Proj4 EPSG Example
  Convert SPCS83 Maryland zone (meters) to Latitude and Longitude

Projection Input:
  Code - CRS: 26985
  CRS Name: NAD83 / Maryland
  CRS Type: projected
  Coord Sys code: 4499
  CS Type: Cartesian
  Dimension: 2
  Remarks: Used in projected and engineering coordinate reference systems.
  CRS Name: NAD83
  Datum Name: North American Datum 1983
  Datum Origin: Origin at geocentre.
  Ellipsoid Name: GRS 1980
  Ellipsoid Unit: metre
  Coord Operation Name: SPCS83 Maryland zone (meters)
  Coord Op Method Name: Lambert Conic Conformal (2SP)

Output
  Unprojected Latitude and Longitude, probably with GRS80 ellipsoids.

Copyright
  Copyright 2007 Michael R. Davis

License
  MIT, BSD, Perl, or GPL


lib/Geo/Proj4.pm
package Geo::Proj4;

use strict;
use warnings;
#our $VERSION = '0.96';

use base 'DynaLoader';

use Scalar::Util   qw/dualvar/;
use Carp           qw/croak/;

# The library definitions
bootstrap Geo::Proj4; # $VERSION;

=chapter NAME

Geo::Proj4 - PROJ.4 cartographic projections library

=chapter SYNOPSIS

  use Geo::Proj4;

  my $proj = Geo::Proj4->new(proj => "merc",
     ellps => "clrk66", lon_0 => -96)
       or die "parameter error: ".Geo::Proj4->error. "\n";

  my $proj = Geo::Proj4->new("+proj=merc +ellps=clrk66 +lon_0=-96")
       or die "parameter error: ".Geo::Proj4->error. "\n";

  my $proj = Geo::Proj4->new(init => "epsg:28992");

  my ($x, $y) = $proj->forward($lat, $lon);

  if($proj->hasInverse)
  {   my ($lat, $lon) = $proj->inverse($x, $y);
      ...
  }

  my $proj = Geo::Proj4->new(init => "epsg:26985") or die;
  my ($lat, $lon) = $proj->inverse(401717.80, 130013.88);

  my $point = [ 123.12, -5.4 ];
  my $projected_point = $from->transform($to, $point);
  my $projected_multi = $from->transform($to, \@points);

=chapter DESCRIPTION

The Open Source PROJ.4 library converts between geographic coordinate
systems.  It is able to convert between geodetic latitude and longitude
(LL, most commonly the WGS84 projection), into an enormous variety of
other cartographic projections (XY, usually UTM).

WARNING: It is not always clear what the source projection is when
M<forward()> or M<inverse()> are used, i.e. in what projection system the
source data is expected to be in.  Therefore, you can better be specific
on both source and destination projection and use M<transform()>.

=chapter METHODS

=section Instantiation

=c_method new $string|%options
The object defines the target projection, but that's easier said than
done: projections have different parameter needs.  The parameters which
can (or need to) be used are listed with C<cs2cs -lP>.  The manual
page of C<cs2cs> explains how the configuration works.

Two ways are provided to define the projection.  Either, use a list
of %options, which are pairs of parameters, or pass one string which
contains all parameters at once.  You must supply a C<proj> parameter.

In case of an OPTION list: WARNING: Specify boolean parameters (e.g. the
south parameter to the UTM projection) with a matching value of undef.

=examples
 my $proj = Geo::Proj4->new(proj => "merc",
    ellps => "clrk66", lon_0 => -96 )
       or die Geo::Proj4->error;

 my $proj = Geo::Proj4->new("+proj=merc +ellps=clrk66 +lon_0=096")
    or die Geo::Proj4->error;

 my $proj = Geo::Proj4->new(init => "epsg:$epsg");
=cut

my $last_error;

sub new($@)
{   my $class = shift;

    my $def;
    if(@_==1)
    {   $def = shift;
    }
    else
    {   my @args;
        while(@_)
        {   my ($key, $val) = (shift, shift);
            push @args, "+$key".(defined $val ? "=$val" : '');
	}
        $def = join ' ', @args;
    }

    my ($self, $error, $errtxt) = new_proj4($def);

    defined $self
        or $last_error = dualvar($error, $errtxt);

    $self;
}

#--------------
=section Accessors

=c_method error
Returns a dualvar (see M<Scalar::Util>) containing the
error number and error string of the last reported error.

=example
 my $proj = Geo::Proj4->new(...);
 unless(defined $proj)
 {   my $error = Geo::Proj4->error;
     warn "error-code: ".$error+0;
     warn "error-string: $error\n";
 }
=cut

sub error() { $last_error }

=method normalized
Returns a string which is produced by the library based on the data
extracted from the initiation parameters.  This string may be more
explicit than the passed values, and could be used for debugging.
=cut

sub normalized()
{   my $norm = normalized_proj4(shift);
    $norm =~ s/^\s+//;
    $norm;
}

=method datum
Tries to return a datum name for this projection.
=cut

sub datum()
{   my $norm = shift->normalized;
    $norm =~ m/\+datum\=(w+)/ ? $1 : undef;
}

=method projection
Returns the projection type.
=cut

sub projection()
{   my $norm = shift->normalized;
    $norm =~ m/\+proj\=(\w+)/ ? $1 : undef;
}

=method dump
Write the definition in extended form to stdout.  This output cannot be
caught, because it is done on stdio level, below the reach of PerlIO.
=cut

sub dump() { dump_proj4(shift) }

=method isLatlong
Returns true when the source projection is using a geodetic coordinate
system; i.e. uses lat long coordinates.  Same as M<isGeodesic()>.

=method isGeodesic
Returns true when the source projection is using a geodetic coordinate
system; i.e. uses lat long coordinates.  Same as M<isLatlong()>
=cut

sub isLatlong()  { is_latlong_proj4(shift) }
sub isGeodesic() { is_latlong_proj4(shift) }

=method isGeocentric
Returns true when the source projection is using a geocentric coordinate
system; i.e. uses x-y coordinates.
=cut

sub isGeocentric() { is_geocentric_proj4(shift) }

=method hasInverse
Returns whether the reverse function for the projection exists.  Some
projections are one-way.
=cut

sub hasInverse() { has_inverse_proj4(shift) }

#--------------
=section Converters

=method forward $latitude, $longitude

Perform a forward projection from $latitude and $longitude (LL) to the
cartographic projection (XY) represented by the Geo::Proj4 instance.

WARNING: for historic reasons, latitude and longitude are assumed to be in 
(floating point) degrees, although the library expects rads.  See
M<forwardRad()>. A latitude south of the Equator and longitude west of
the Prime Meridian given with negative values.

Returned are two values, usually X and Y in meters, or whatever units are
relevant to the given projection.  When the destination projection also
than the order of parameters will be returned as LONG,LAT (not lat,long!)

On error, C<forward> will return undef for both values.

=example
 my ($x, $y) = $proj->forward($lat, $lon);
 my ($long2, $lat2) = $proj->forward($lat, $lon);

=cut

sub forward($$)
{   my ($self, $lat, $long) = @_;
    forward_degrees_proj4($self, $lat, $long);
}

=method forwardRad $latitude, $longitude

Perform a forward projection from $latitude and $longitude (LL) to the
cartographic projection (XY) represented by the Geo::Proj4 instance.
This function reflects to library function C<forward()>, expecting
radians, not degrees.

=cut

sub forwardRad($$)
{   my ($self, $lat, $long) = @_;
    forward_proj4($self, $lat, $long);
}

=method inverse ($x,$y) | ($lat,$long)

Perform an inverse projection from the (cartographic) projection represented
by this Geo::Proj4 object, back into latitude and longitude values.

WARNING: for historic reasons, latitude and longitude are assumed to be in 
(floating point) degrees, although the library expects rads.  See
M<inverseRad()>.

On error, C<inverse> will return undef for both values.

=example

  if($proj->hasInverse)
  {  my ($lat, $lon) = $proj->inverse($x, $y);
     ...
  }

=cut

sub inverse($$) { inverse_degrees_proj4(@_) }

=method inverseRad ($x,$y) | ($lat|$long)

Perform an inverse projection from the (cartographic) projection
represented by this Geo::Proj4 object, back into latitude and longitude
values.  Latitude and longitude are assumed to be in radians. See
M<inverse()>.
=cut

sub inverseRad($$) { inverse_proj4(@_) }

=method transform $to, $point|ARRAY-of-$points
Translate the $points into the projecten of $to.  Each point is specified
as two or three values in an ARRAY.  In case of latlong source or
destination projections, coordinates are translated into radians and/or
back.  Both input and output values are always in X-Y/LongLat order.
See M<transformRad()>

=example
 my $from  = Geo::Proj4->new("+proj=latlong +datum=NAD83");
 my $to    = Geo::Proj4->new("+proj=utm +zone=10 +datum=WGS84");

 my $point = [ 1.12, 3.25 ];  # See Geo::Point
 my $pr_point = $from->transform($to, $point);

 my $pr    = $from->transform($to, [ $point1, $point2 ]);
 my $pr_point1 = $pr->[0];
 my $pr_point2 = $pr->[1];

=error transform() expects array of points
=cut

sub transform($$)
{   my ($self, $to, $points) = @_;

    ref $points eq 'ARRAY'
        or croak "ERROR: transform() expects array of points";

    my ($err, $errtxt, $pr);
    if(ref($points->[0]) eq 'ARRAY')
    {   ($err, $errtxt, $pr) = transform_proj4($self, $to, $points, 1);
    }
    else
    {   ($err, $errtxt, $pr) = transform_proj4($self, $to, [$points], 1);
        $pr = $pr->[0] if $pr;
    }

    $last_error = dualvar $err, $errtxt;
    $err ? () : $pr;
}

=method transformRad $to, $point|ARRAY-of-$points
Translate the $points into the projecten of $to.  Each point is specified
as two or three values in an ARRAY.  In case of latlong source or
destination projections, coordinates are expected to be in radians.
Both input and output values are always in X-Y/LongLat order.
See M<transform()>

=error transformRad() expects array of points
=cut

sub transformRad($$)
{   my ($self, $to, $points) = @_;

    ref $points eq 'ARRAY'
        or croak "ERROR: transformRad() expects array of points";

    my ($err, $errtxt, $pr);
    if(ref($points->[0]) eq 'ARRAY')
    {   ($err, $errtxt, $pr) = transform_proj4($self, $to, $points, 0);
    }
    else
    {   ($err, $errtxt, $pr) = transform_proj4($self, $to, [$points], 0);
        $pr = $pr->[0] if $pr;
    }

    $last_error = dualvar $err, $errtxt;
    $err ? () : $pr;
}

sub AUTOLOAD(@)
{   our $AUTOLOAD;
    die "$AUTOLOAD not implemented";
}

#--------------
=section Library introspection

=ci_method libVersion
Returns the version of the proj4 library
=cut

sub libVersion()
{   my $version = libproj_version_proj4();
    $version =~ s/./$&./g;
    $version;
}

=c_method listTypes
Returns a list with all defined projection types.

=example
 foreach my $id (Geo::Proj4->listTypes)
 {   my $def = Geo::Proj4->type($id);
     print "$id = $def->{description}\n";
 }
=cut

sub listTypes() { &def_types_proj4 }

=c_method typeInfo $label
Returns a hash with information about the specified projection type.  With
M<listTypes()>, all defined LABELS can be found.
=cut

sub typeInfo($)
{   my $label = $_[1];
    my %def = (id => $label);
    my($descr) = type_proj4($label);
    $def{has_inverse} = not ($descr =~ s/(?:\,?\s+no\s+inv\.?)//);
    $def{description} = $descr;
    \%def;
}

=c_method listEllipsoids
Returns a list with all defined ellips labels.

=example
 foreach my $id (Geo::Proj4->listEllipsoids)
 {   my $def = Geo::Proj4->ellipsoid($id);
     print "$id = $def->{name}\n";
 }
=cut

sub listEllipsoids() { &def_ellps_proj4 }

=c_method ellipsoidInfo $label
Returns a hash with information about the specified ellipsis.  With
M<listEllipsoids()>, all defined LABELS can be found.
=cut

sub ellipsoidInfo($)
{   my $label = $_[1];
    my %def = (id => $label);
    @def{ qw/major ell name/ } = ellps_proj4($label);
    \%def;
}

=c_method listUnits
Returns a list with all defined unit labels.

=example
 foreach my $id (Geo::Proj4->listUnits)
 {   my $def = Geo::Proj4->unit($id);
     print "$id = $def->{name}\n";
 }
=cut

sub listUnits() { &def_units_proj4 }

=c_method unitInfo $label
Returns a hash with information about the specified unit.  With
M<listUnits()>, all defined LABELS can be found.
=cut

sub unitInfo($)
{   my $label = $_[1];
    my %def = (id => $label);
    @def{ qw/to_meter name/ } = unit_proj4($label);
    $def{to_meter} =~ s!^1\.?/(.*)!1/$1!e;  # 1/2 -> 0.5
    \%def;
}

=c_method listDatums
Returns a list with all defined datum labels.

=example
 foreach my $id (Geo::Proj4->listDatums)
 {   my $def = Geo::Proj4->datum($id);
     print "$id = $def->{ellips_id}\n";
 }
=cut

sub listDatums() { &def_datums_proj4 }

=c_method datumInfo $label
Returns a hash with information about the specified datum.  With
M<listDatums()>, all defined LABELS can be found.
=cut

sub datumInfo($)
{   my $label = $_[1];
    my %def = (id => $label);
    @def{ qw/ellipse_id definition comments/ } = datum_proj4($label);
    \%def;
}

#--------------
=chapter DETAILS

=section Install

Geo::Proj4 uses XS to wrap the PROJ.4 cartographic projections library. You
will need to have the PROJ.4 library installed in order to build and use
this module. You can get source code and binaries for the PROJ.4 library
from its home page at L<http://www.remotesensing.org/proj/>.

=section Projections

Covering all the possible projections and their arguments in PROJ.4
is well beyond the scope of this document. However, the C<cs2cs(1)>
utility that ships with PROJ.4 will list the projections it knows about
by running B<cs2cs -lp>, the ellipsoid models it knows with the B<-le>
parameter, the units it knows about with B<-lu>, and the geodetic datums
it knows with B<-ld>. Read L<cs2cs(1)> for more details.

Alternately, you can read the PROJ.4 documentation, which can be found
on the project's homepage. There are links to PDFs, text documentation,
a FAQ, and more.

=section Bugs

One common source of errors is that latitude and longitude are
swapped: some projection systems use lat-long, other use x-y
which is a swapped order.  Especially the M<forward()> and M<inverse()>
cause this problem, always flipping the coordinate order.  The
M<transform()> method is much easier: input and output in x-y/long-lat
order.

Also be warned  that the values must have the right sign. Make sure you
give negative values for south latitude and west longitude.  For
calculating projections, this is more important than on maps.

=cut

# more text in PODTAIL.txt

1;
